PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp

```diff
@@ -534,6 +743,259 @@ void ProjectPanel::openSelectFile()
+    hChildItem = _treeView.getNextSibling(hChildItem))
+    {
+		itemVisibilityChanges(hChildItem, visible);
+	}
+
+}
+
+void ProjectPanel::expandOrCollapseDirectory(bool expand, HTREEITEM hItem)
+{
+
+	TVITEM tvItem;
+	tvItem.mask = TVIF_PARAM;
+	tvItem.hItem = hItem;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
+	if (!(data.isDirectory() || data.isBaseDirectory()) )
+		return;
+
+	if (expand)
+	{
+		removeDummies(hItem);
+		if (data.isDirectory())
+			data.watchDir(true);
+	}
+	else
+	{
+		if (data.isDirectory())
+			data.watchDir(false);
+	}
+
+	for (HTREEITEM hChildItem = _treeView.getChildFrom(hItem);
+    hChildItem != NULL;
+    hChildItem = _treeView.getNextSibling(hChildItem))
+    {
+		itemVisibilityChanges(hChildItem, expand);
+	}
+
+	_directoryWatcher.update();
+
+}
+
+bool ProjectPanel::setFilters(const std::vector<generic_string>& filters, HTREEITEM hItem)
+{
+	TVITEM tvItem;
+	tvItem.mask = TVIF_PARAM;
+	tvItem.hItem = hItem;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
+
+	if (data.isDirectory() || data.isBaseDirectory())
+	{
+		if (data.isBaseDirectory() && data._filters == filters)
+			return false;
+
+		data._filters = filters;
+	}
+
+	if (data.isWatching())
+	{
+		data.watchDir(true);
+	}
+
+	for (HTREEITEM hItemNode = _treeView.getChildFrom(hItem);
+		hItemNode != NULL;
+		hItemNode = _treeView.getNextSibling(hItemNode))
+	{
+		setFilters(filters,hItemNode);
+	}
+
+	updateBaseDirName(hItem);
+	return true;
+}
+
+void ProjectPanel::updateBaseDirName(HTREEITEM hItem)
+{
+	TCHAR textBuffer[MAX_PATH];
+	TVITEM tvItem;
+	tvItem.mask = TVIF_TEXT | TVIF_PARAM;
+	tvItem.pszText = textBuffer;
+	tvItem.cchTextMax = MAX_PATH;
+	tvItem.hItem = hItem;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
+
+	if (data.isDirectory() || data.isDirectoryFile())
+	{
+		updateBaseDirName(_treeView.getParent(hItem));
+	}
+	else if (data.isBaseDirectory())
+	{
+		if (data._userLabel.empty())
+		{
+			generic_string newFolderLabel = buildDirectoryName(data._filePath, data._filters);
+			wcsncpy(tvItem.pszText,newFolderLabel.c_str(), MAX_PATH-1);
+			tvItem.mask = TVIF_TEXT;
+			::SendMessage(_treeView.getHSelf(), TVM_SETITEM, 0,reinterpret_cast<LPARAM>(&tvItem));
+		}
+	}
+}
+
+void ProjectPanel::removeDummies(HTREEITEM hTreeItem)
+{
+	HTREEITEM hItemNode = _treeView.getChildFrom(hTreeItem);
+	if (hItemNode == NULL)
+		return;
+
+	TVITEM tvItem;
+	tvItem.mask = TVIF_PARAM;
+	tvItem.hItem = hItemNode;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);

```

Comment:
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.